miner/scroll_worker.go

```diff
@@ -0,0 +1,943 @@
+		if res != nil && res.FinalBlock != nil {
+			w.updateSnapshot(res.FinalBlock)
+		}
+		w.currentPipeline = nil
+		return nil
+	}
+
+	if res == nil || res.FinalBlock == nil {
+		w.startNewPipeline(time.Now().Unix())
+		return nil
+	}
+	return w.commit(res)
+}
+
+// commit runs any post-transaction state modifications, assembles the final block
+// and commits new work if consensus engine is running.
+func (w *worker) commit(res *pipeline.Result) error {
+	defer func(t0 time.Time) {
+		l2CommitTimer.Update(time.Since(t0))
+	}(time.Now())
+
+	if res.CCCErr != nil {
+		commitReasonCCCCounter.Inc(1)
+	} else {
+		commitReasonDeadlineCounter.Inc(1)
+	}
+
+	block, err := w.engine.FinalizeAndAssemble(w.chain, res.FinalBlock.Header, res.FinalBlock.State,
+		res.FinalBlock.Txs, nil, res.FinalBlock.Receipts)
+	if err != nil {
+		return err
+	}
+
+	select {
+	case w.taskCh <- &task{receipts: res.FinalBlock.Receipts, state: res.FinalBlock.State, block: block, createdAt: time.Now(),
+		accRows: res.Rows, nextL1MsgIndex: res.FinalBlock.NextL1MsgIndex}:
+		w.unconfirmed.Shift(block.NumberU64() - 1)
+		log.Info("Commit new mining work", "number", block.Number(), "sealhash", w.engine.SealHash(block.Header()),
+			"txs", res.FinalBlock.Txs.Len(),
+			"gas", block.GasUsed(), "fees", totalFees(block, res.FinalBlock.Receipts),
+			"elapsed", common.PrettyDuration(time.Since(w.currentPipelineStart)))
+	case <-w.exitCh:
+		log.Info("Worker has exited")
+	}
+
+	w.currentPipeline = nil
+	return nil
+}
+
+// copyReceipts makes a deep copy of the given receipts.
+func copyReceipts(receipts []*types.Receipt) []*types.Receipt {
+	result := make([]*types.Receipt, len(receipts))
+	for i, l := range receipts {
+		cpy := *l
+		result[i] = &cpy
+	}
+	return result
+}
+
+// postSideBlock fires a side chain event, only use it for testing.
+func (w *worker) postSideBlock(event core.ChainSideEvent) {
+	select {
+	case w.chainSideCh <- event:
+	case <-w.exitCh:
+	}
+}
+
+func (w *worker) onTxFailingInPipeline(txIndex int, tx *types.Transaction, err error) bool {
+	writeTrace := func() {
+		var trace *types.BlockTrace
+		var errWithTrace *pipeline.ErrorWithTrace
+		if w.config.StoreSkippedTxTraces && errors.As(err, &errWithTrace) {
+			trace = errWithTrace.Trace
+		}
+		rawdb.WriteSkippedTransaction(w.eth.ChainDb(), tx, trace, err.Error(),
+			w.currentPipeline.Header.Number.Uint64(), nil)
+	}
+
+	switch {
+	case errors.Is(err, core.ErrGasLimitReached) && tx.IsL1MessageTx():
+		// If this block already contains some L1 messages,
+		// terminate here and try again in the next block.
+		if txIndex > 0 {
+			break
+		}
+		// A single L1 message leads to out-of-gas. Skip it.
+		queueIndex := tx.AsL1MessageTx().QueueIndex
+		log.Info("Skipping L1 message", "queueIndex", queueIndex, "tx", tx.Hash().String(), "block",
+			w.currentPipeline.Header.Number, "reason", "gas limit exceeded")
+		writeTrace()
+		l1TxGasLimitExceededCounter.Inc(1)
+
+	case errors.Is(err, core.ErrInsufficientFunds):
```

Comment:
there's a reason why we did unwrap before 
(`errors.Is(err, core.ErrInsufficientFunds) || errors.Is(errors.Unwrap(err), core.ErrInsufficientFunds)`)

https://github.com/scroll-tech/go-ethereum/blob/develop/core/state_transition.go#L256

Reply:
`errors.Is` already handles wrapped errors. https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/errors/wrap.go;l=29-31